![tracker](https://us-central1-vertex-ai-mlops-369716.cloudfunctions.net/pixel-tracking?path=statmike%2Fvertex-ai-mlops%2Fdata%2Bai%2Fbq-ml%2Fmodels%2Flinear_regression&file=linear_regression.ipynb)
<!--- header table --->
<table align="left">
<tr>
  <td style="text-align: center">
    <a href="https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/models/linear_regression/linear_regression.ipynb">
      <img width="32px" src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub logo">
      <br>View on<br>GitHub
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/models/linear_regression/linear_regression.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo">
      <br>Run in<br>Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https%3A%2F%2Fraw.githubusercontent.com%2Fstatmike%2Fvertex-ai-mlops%2Fmain%2Fdata%252Bai%2Fbq-ml%2Fmodels%2Flinear_regression%2Flinear_regression.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo">
      <br>Run in<br>Colab Enterprise
    </a>
  </td>      
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/bigquery/import?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/models/linear_regression/linear_regression.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/bigquery/v1/32px.svg" alt="BigQuery logo">
      <br>Open in<br>BigQuery Studio
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/models/linear_regression/linear_regression.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo">
      <br>Open in<br>Vertex AI Workbench
    </a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Share This On: </b> 
    <a href="https://www.linkedin.com/sharing/share-offsite/?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/models/linear_regression/linear_regression.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a> 
    <a href="https://reddit.com/submit?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/models/linear_regression/linear_regression.ipynb"><img src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit Logo" width="20px"></a> 
    <a href="https://bsky.app/intent/compose?text=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/models/linear_regression/linear_regression.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://twitter.com/intent/tweet?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/models/linear_regression/linear_regression.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a> 
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Connect With Author On: </b> 
    <a href="https://www.linkedin.com/in/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a>
    <a href="https://www.github.com/statmike"><img src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub Logo" width="20px"></a> 
    <a href="https://www.youtube.com/@statmike-channel"><img src="https://upload.wikimedia.org/wikipedia/commons/f/fd/YouTube_full-color_icon_%282024%29.svg" alt="YouTube Logo" width="20px"></a>
    <a href="https://bsky.app/profile/statmike.bsky.social"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://x.com/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/models/linear_regression/linear_regression.ipynb"><img src="https://www.svgrepo.com/download/5445/download-button.svg" alt="Download icon" width="20px"></a> <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/models/linear_regression/linear_regression.ipynb">Download Notebook</a> <i>(right-click and "Save As")</i>
  </td>
</tr>
</table>

# Linear Regression — BigQuery ML

Train a **regression** model entirely in SQL with `CREATE MODEL`, then walk the full model lifecycle: evaluate, predict, explain, inspect coefficients, apply in-model preprocessing, and tune hyperparameters. No data leaves BigQuery, and no separate training infrastructure is needed.

**Lifecycle:** `CREATE MODEL` → `ML.EVALUATE` → `ML.PREDICT` → `ML.EXPLAIN_PREDICT` → `ML.GLOBAL_EXPLAIN` → `ML.WEIGHTS` → `ML.FEATURE_INFO` / `ML.TRAINING_INFO` → `TRANSFORM` clause → hyperparameter tuning (`ML.TRIAL_INFO`)

**When to use linear regression:**
- Predicting a continuous numeric value (price, weight, demand) from features
- A fast, explainable baseline before trying boosted trees or DNNs
- You want directly interpretable coefficients (`ML.WEIGHTS`) plus per-prediction and global feature attributions
- The workhorse behind regression-based forecasting (lagged-feature design matrices)

**Data:** [`bigquery-public-data.ml_datasets.penguins`](https://console.cloud.google.com/marketplace/product/bigquery-public-datasets) — predict a penguin's `body_mass_g` from species, island, and bill/flipper measurements. Rows with a NULL label or an unrecorded `sex` are filtered out.

**Featured in:** [Regression-Based Forecasting](../../workflows/regression_based_forecasting/) — time/lag/lead feature engineering applied to `LINEAR_REG` for demand forecasting.

**References:** [Full reference](../../RESOURCES.md) | [CREATE MODEL (GLM) docs](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-create-glm) | [Setup guide](../../setup/)

---
## Setup

Set your project and location, authenticate, and create a shared dataset for the model.

> Linear regression trains on data already in BigQuery — no connection or remote model required. The model is created with `CREATE MODEL` and stored in your dataset. See the [Setup Reference](../../setup/) for details.

In [1]:
PROJECT_ID = 'statmike-mlops-349915'  # <-- Replace with your project ID
LOCATION = 'US'  # BigQuery dataset location
DATASET_ID = 'bq_ml'  # Shared dataset across all bq-ml notebooks

### Environment

> **Already set up the project environment?** The cell below is a no-op — packages are already in your kernel. See the [Setup Reference](../../setup/) for details.
>
> **Running standalone** (Colab, Colab Enterprise, Vertex AI Workbench)? The cell below installs required packages into your current kernel.

In [2]:
import subprocess, sys, shutil

def install(*packages):
    """Install packages using uv (fast) with pip fallback."""
    uv = shutil.which('uv')
    if uv:
        subprocess.check_call([uv, 'pip', 'install', '-q', '--python', sys.executable, *packages])
    else:
        subprocess.check_call([sys.executable, '-m', 'pip', 'install', '-q', '--upgrade', *packages])

install('google-cloud-bigquery', 'db-dtypes', 'bigquery-magics', 'bigframes', 'matplotlib')

In [3]:
# Authenticate (Colab only — other environments use Application Default Credentials)
try:
    from google.colab import auth
    auth.authenticate_user()
except ImportError:
    pass  # Not in Colab — ADC is used automatically

In [4]:
from google.cloud import bigquery
import pandas as pd

client = bigquery.Client(project=PROJECT_ID)
pd.set_option('display.max_colwidth', None)

# Create the shared dataset (idempotent)
dataset_ref = bigquery.DatasetReference(PROJECT_ID, DATASET_ID)
dataset = bigquery.Dataset(dataset_ref)
dataset.location = LOCATION
client.create_dataset(dataset, exists_ok=True)
print(f'Dataset {PROJECT_ID}.{DATASET_ID} ready')

# Register %%bigquery cell magic (auto-loaded in Colab, needed elsewhere)
%load_ext bigquery_magics

Dataset statmike-mlops-349915.bq_ml ready


---
## Step 1 — Create the model with `CREATE MODEL`

`CREATE MODEL` trains and stores the model in your dataset. The essentials are `model_type` and `input_label_cols`. We also:
- `data_split_method = 'AUTO_SPLIT'` — automatically hold out rows for evaluation
- `category_encoding_method = 'DUMMY_ENCODING'` — drop one baseline category per categorical feature instead of one-hot-encoding all of them. This matters for `ML.WEIGHTS` in Step 5: with the default `ONE_HOT_ENCODING`, categorical dummies plus the intercept are collinear, so individual `category_weights` are **not uniquely identified** — re-running `CREATE MODEL` with a different random split can swing a category's weight by thousands of grams, even though predictions stay stable. `DUMMY_ENCODING` fixes this.
- `enable_global_explain = TRUE` — **required** to use `ML.GLOBAL_EXPLAIN` later

Training runs synchronously — the cell completes when the model is ready.

> For a small, unregularized problem like this one, BigQuery ML auto-selects the `NORMAL_EQUATION` solver (a single closed-form pass) instead of iterative gradient descent — see the `ML.TRAINING_INFO` note in Step 6.
- `calculate_p_values = TRUE` and `l1_reg = 0` — these unlock `ML.ADVANCED_WEIGHTS` in Step 6, which reports a standard error and p-value for every coefficient. Both must be set **here**, at `CREATE MODEL` time; p-values cannot be added to an already-trained model. `l1_reg = 0` is the default, but it is stated explicitly so the dependency is visible.

The cell below emits a warning — *"Since model contains categorical values, regression statistics will not be calculated for unstandardized intercept."* That is expected. Step 6 shows how to get intercept statistics anyway.


In [5]:
query = f"""
CREATE OR REPLACE MODEL `{PROJECT_ID}.{DATASET_ID}.linear_regression_penguins`
OPTIONS(
  model_type = 'LINEAR_REG',
  input_label_cols = ['body_mass_g'],
  data_split_method = 'AUTO_SPLIT',
  category_encoding_method = 'DUMMY_ENCODING',
  calculate_p_values = TRUE,
  l1_reg = 0,
  enable_global_explain = TRUE
) AS
SELECT
  species, island, culmen_length_mm, culmen_depth_mm, flipper_length_mm, sex, body_mass_g
FROM `bigquery-public-data.ml_datasets.penguins`
WHERE body_mass_g IS NOT NULL AND sex IN ('MALE', 'FEMALE')
"""
client.query(query).result()
print('Model linear_regression_penguins created')

Model linear_regression_penguins created


---
## Step 2 — Evaluate with `ML.EVALUATE`

`ML.EVALUATE` returns standard regression metrics on the automatically held-out evaluation split: mean absolute error, mean squared error, R², and explained variance.

In [6]:
query = f"""
SELECT *
FROM ML.EVALUATE(MODEL `{PROJECT_ID}.{DATASET_ID}.linear_regression_penguins`)
"""
client.query(query).to_dataframe()

,mean_absolute_error,mean_squared_error,mean_squared_log_error,median_absolute_error,r2_score,explained_variance
0,226.413652,80658.955707,0.004982,173.534936,0.875223,0.875223


---
## Step 3 — Predict with `ML.PREDICT`

`ML.PREDICT` scores new rows. It returns `predicted_body_mass_g` — a single continuous value, no probability array like a classifier. You pass any query with the same feature columns.

In [7]:
query = f"""
SELECT
  predicted_body_mass_g,
  species,
  flipper_length_mm
FROM ML.PREDICT(
  MODEL `{PROJECT_ID}.{DATASET_ID}.linear_regression_penguins`,
  (SELECT * FROM `bigquery-public-data.ml_datasets.penguins`
   WHERE body_mass_g IS NOT NULL AND sex IN ('MALE', 'FEMALE') LIMIT 10)
)
"""
client.query(query).to_dataframe()

,predicted_body_mass_g,species,flipper_length_mm
0,3383.869817,Adelie Penguin (Pygoscelis adeliae),184.0
1,3876.602681,Adelie Penguin (Pygoscelis adeliae),184.0
2,3883.095842,Adelie Penguin (Pygoscelis adeliae),184.0
3,3399.758077,Chinstrap penguin (Pygoscelis antarctica),192.0
4,3418.389700,Adelie Penguin (Pygoscelis adeliae),192.0
5,4027.809144,Adelie Penguin (Pygoscelis adeliae),192.0
6,3319.185745,Chinstrap penguin (Pygoscelis antarctica),192.0
7,3636.211101,Chinstrap penguin (Pygoscelis antarctica),200.0
8,4045.791299,Chinstrap penguin (Pygoscelis antarctica),200.0
9,4211.269935,Adelie Penguin (Pygoscelis adeliae),200.0


---
## Step 4 — Explain predictions

BigQuery ML provides explainability out of the box:
- **`ML.EXPLAIN_PREDICT`** — per-row feature attributions (which features pushed *this* prediction up or down)
- **`ML.GLOBAL_EXPLAIN`** — overall feature importance across the model (requires `enable_global_explain = TRUE`)

In [8]:
query = f"""
SELECT
  predicted_body_mass_g,
  top_feature_attributions
FROM ML.EXPLAIN_PREDICT(
  MODEL `{PROJECT_ID}.{DATASET_ID}.linear_regression_penguins`,
  (SELECT * FROM `bigquery-public-data.ml_datasets.penguins`
   WHERE body_mass_g IS NOT NULL AND sex IN ('MALE', 'FEMALE') LIMIT 5),
  STRUCT(5 AS top_k_features)
)
"""
client.query(query).to_dataframe()

,predicted_body_mass_g,top_feature_attributions
0,3383.869817,"[{'feature': 'sex', 'attribution': -387.224255296456}, {'feature': 'flipper_length_mm', 'attribution': -275.5181977215216}, {'feature': 'culmen_length_mm', 'attribution': -134.46983749530466}, {'feature': 'culmen_depth_mm', 'attribution': 83.46478686274807}, {'feature': 'island', 'attribution': -13.103051308902195}]"
1,3876.602681,"[{'feature': 'flipper_length_mm', 'attribution': -275.5181977215216}, {'feature': 'culmen_depth_mm', 'attribution': 130.76758729480994}, {'feature': 'culmen_length_mm', 'attribution': -76.26402921071723}, {'feature': 'island', 'attribution': -13.103051308902195}, {'feature': 'species', 'attribution': 0.0}]"
2,3883.095842,"[{'feature': 'flipper_length_mm', 'attribution': -275.5181977215216}, {'feature': 'culmen_depth_mm', 'attribution': 117.25250145707784}, {'feature': 'culmen_length_mm', 'attribution': -56.255782612890236}, {'feature': 'island', 'attribution': -13.103051308902195}, {'feature': 'species', 'attribution': 0.0}]"
3,3399.758077,"[{'feature': 'sex', 'attribution': -387.224255296456}, {'feature': 'species', 'attribution': -260.30629455232906}, {'feature': 'flipper_length_mm', 'attribution': -145.61014838875457}, {'feature': 'culmen_depth_mm', 'attribution': 49.677072268418286}, {'feature': 'culmen_length_mm', 'attribution': 45.60438188513792}]"
4,3418.389700,"[{'feature': 'sex', 'attribution': -387.224255296456}, {'feature': 'flipper_length_mm', 'attribution': -145.61014838875457}, {'feature': 'culmen_length_mm', 'attribution': -121.7373169330512}, {'feature': 'culmen_depth_mm', 'attribution': -24.655899839107075}, {'feature': 'island', 'attribution': -13.103051308902195}]"


In [9]:
query = f"""
SELECT *
FROM ML.GLOBAL_EXPLAIN(MODEL `{PROJECT_ID}.{DATASET_ID}.linear_regression_penguins`)
ORDER BY attribution DESC
"""
client.query(query).to_dataframe()

,feature,attribution
0,species,406.139459
1,flipper_length_mm,197.569151
2,sex,191.867874
3,culmen_depth_mm,112.038866
4,culmen_length_mm,85.859295
5,island,11.623621


---
## Step 5 — Inspect coefficients with `ML.WEIGHTS`

Unlike a classifier, a `LINEAR_REG` model's weights are directly interpretable in the label's own units — e.g. "each additional mm of flipper length adds N grams of predicted body mass." Numeric features get a single `weight`; categorical features (like `species`, `island`, `sex`) expand into a `category_weights` array. Because the model was trained with `category_encoding_method = 'DUMMY_ENCODING'`, one category per feature is the pinned baseline (`weight: 0.0`) and every other category's weight is a stable, well-defined delta from it. This function only applies to `LINEAR_REG` / `LOGISTIC_REG` / `MATRIX_FACTORIZATION` models.

In [10]:
query = f"""
SELECT *
FROM ML.WEIGHTS(MODEL `{PROJECT_ID}.{DATASET_ID}.linear_regression_penguins`)
"""
client.query(query).to_dataframe()

,processed_input,weight,category_weights
0,species,NaN,"[{'category': 'Adelie Penguin (Pygoscelis adeliae)', 'weight': 0.0}, {'category': 'Chinstrap penguin (Pygoscelis antarctica)', 'weight': -260.30629455232906}, {'category': 'Gentoo penguin (Pygoscelis papua)', 'weight': 987.7614450022897}]"
1,island,NaN,"[{'category': 'Biscoe', 'weight': 0.0}, {'category': 'Dream', 'weight': -13.103051308902195}, {'category': 'Torgersen', 'weight': -48.063625808092205}]"
2,culmen_length_mm,18.189315,[]
3,culmen_depth_mm,67.575429,[]
4,flipper_length_mm,16.238506,[]
5,sex,NaN,"[{'category': 'MALE', 'weight': 0.0}, {'category': 'FEMALE', 'weight': -387.224255296456}]"
6,__INTERCEPT__,-1112.804840,[]


---
## Step 6 — Test the coefficients with `ML.ADVANCED_WEIGHTS`

`ML.WEIGHTS` (Step 5) answers *"how big is this coefficient?"* — `ML.ADVANCED_WEIGHTS` answers *"is it distinguishable from zero at all?"*

It also flattens the shape. Where `ML.WEIGHTS` puts numeric features in `weight` and nests categoricals in a `category_weights` array, `ML.ADVANCED_WEIGHTS` gives **one row per coefficient** — `processed_input`, `category`, `weight`, `standard_error`, `p_value` — plus a final `__INTERCEPT__` row.

Requirements, all set back in Step 1: `calculate_p_values = TRUE`, `category_encoding_method = 'DUMMY_ENCODING'`, and `l1_reg = 0`. Total feature cardinality must also be under 1,000, and only `LINEAR_REG` and **binary** `LOGISTIC_REG` are supported.


In [11]:
query = f"""
SELECT
  processed_input,
  category,
  ROUND(weight, 4) AS weight,
  ROUND(standard_error, 4) AS standard_error,
  p_value
FROM ML.ADVANCED_WEIGHTS(MODEL `{PROJECT_ID}.{DATASET_ID}.linear_regression_penguins`)
"""
client.query(query).to_dataframe()

,processed_input,category,weight,standard_error,p_value
0,species,Adelie Penguin (Pygoscelis adeliae),0.0000,0.0000,NaN
1,species,Chinstrap penguin (Pygoscelis antarctica),-260.3063,87.3458,3.341759e-03
2,species,Gentoo penguin (Pygoscelis papua),987.7614,135.3708,5.939470e-08
3,island,Biscoe,0.0000,0.0000,NaN
4,island,Dream,-13.1031,57.7442,8.206483e-01
5,island,Torgersen,-48.0636,60.0926,4.239188e-01
6,culmen_length_mm,None,18.1893,7.0393,1.010192e-02
7,culmen_depth_mm,None,67.5754,19.5516,8.496620e-04
8,flipper_length_mm,None,16.2385,2.8995,2.565310e-06
9,sex,MALE,0.0000,0.0000,NaN


Read the `p_value` column and one feature stands out: **`island` is the only feature whose categories are not significant** (roughly 0.82 for Dream and 0.42 for Torgersen, against 0.05). Once `species` is in the model, where a penguin was found adds nothing — Gentoo appears only on Biscoe and Chinstrap only on Dream, so `island` is largely a restatement of `species`. `ML.WEIGHTS` shows non-zero island weights and gives you no way to notice this.

Two things to watch in the output:

- The dropped `DUMMY_ENCODING` baseline is the **most frequent** category, not the first alphabetically — here Adelie (146 rows), Biscoe (163), and `MALE` (168, beating the alphabetically-earlier `FEMALE`). Baseline rows report `weight` 0.0 and `standard_error` 0.0 with `p_value` `NaN`. They are placeholders, not estimates, so filter them with `IS_NAN(p_value)` rather than reading 0.0 as a measured standard error.
- `__INTERCEPT__` comes back with a **`NULL`** standard error and p-value whenever the model has categorical features — the warning from Step 1. Passing `STRUCT(TRUE AS standardize)` gets them back.


In [12]:
query = f"""
SELECT
  processed_input,
  ROUND(weight, 4) AS weight,
  ROUND(standard_error, 4) AS standard_error,
  p_value
FROM ML.ADVANCED_WEIGHTS(
  MODEL `{PROJECT_ID}.{DATASET_ID}.linear_regression_penguins`,
  STRUCT(TRUE AS standardize)
)
WHERE category IS NULL
"""
client.query(query).to_dataframe()

,processed_input,weight,standard_error,p_value
0,culmen_length_mm,99.4713,38.4955,0.010102
1,culmen_depth_mm,133.0719,38.5017,0.000850
2,flipper_length_mm,227.5951,40.6382,0.000003
3,__INTERCEPT__,4110.7204,71.0032,0.000000


Standardizing rescales each numeric feature to zero mean and unit variance before reporting, and that makes intercept statistics available.

Note what does **not** change: the p-values are identical either way. Standardizing scales `weight` and `standard_error` by the same factor, so the t-statistic — and therefore significance — is unchanged. Standardize to compare effect sizes across features on a common scale (`flipper_length_mm` goes from ~16 grams *per mm* to ~228 grams *per standard deviation*), not to change which features count as significant.

One caution: the standardized intercept is not the average label. It lands near 4111 while mean `body_mass_g` is about 4207, because categorical features stay dummy-coded rather than centered — so the intercept is the prediction at mean numeric features **and** baseline categories, i.e. an Adelie male on Biscoe.


---
## Step 7 — Introspect the model

`ML.FEATURE_INFO` reports the statistics each feature had during training. `ML.TRAINING_INFO` returns the per-iteration loss curve — for this model, BigQuery ML auto-selects the `NORMAL_EQUATION` solver, which trains in a single closed-form pass, so expect exactly one row with no `eval_loss`.

In [13]:
query = f"""
SELECT *
FROM ML.FEATURE_INFO(MODEL `{PROJECT_ID}.{DATASET_ID}.linear_regression_penguins`)
"""
client.query(query).to_dataframe()

,input,min,max,mean,median,stddev,category_count,null_count,dimension
0,species,NaN,NaN,NaN,NaN,NaN,3,0,<NA>
1,island,NaN,NaN,NaN,NaN,NaN,3,0,<NA>
2,culmen_length_mm,32.1,59.6,43.992793,44.1,5.468668,<NA>,0,<NA>
3,culmen_depth_mm,13.1,21.5,17.164865,17.3,1.969235,<NA>,0,<NA>
4,flipper_length_mm,172.0,231.0,200.966967,197.0,14.015765,<NA>,0,<NA>
5,sex,NaN,NaN,NaN,NaN,NaN,2,0,<NA>


In [14]:
query = f"""
SELECT iteration, loss, eval_loss, learning_rate, duration_ms
FROM ML.TRAINING_INFO(MODEL `{PROJECT_ID}.{DATASET_ID}.linear_regression_penguins`)
ORDER BY iteration
"""
client.query(query).to_dataframe()

,iteration,loss,eval_loss,learning_rate,duration_ms
0,0,80658.955707,NaN,NaN,2152


---
## Step 8 — In-model preprocessing with the `TRANSFORM` clause

The `TRANSFORM` clause bakes preprocessing into the model. Whatever you do in `TRANSFORM` (scaling, bucketizing, feature crosses) is **saved with the model and reapplied automatically at predict time** — so `ML.PREDICT` takes raw data, with no need to repeat the preprocessing.

Here we standard-scale the numeric bill/flipper measurements. Notice the prediction query passes raw, unscaled rows.

In [15]:
query = f"""
CREATE OR REPLACE MODEL `{PROJECT_ID}.{DATASET_ID}.linear_regression_penguins_transform`
TRANSFORM(
  ML.STANDARD_SCALER(culmen_length_mm) OVER() AS culmen_length_mm,
  ML.STANDARD_SCALER(flipper_length_mm) OVER() AS flipper_length_mm,
  species, island, sex, body_mass_g
)
OPTIONS(
  model_type = 'LINEAR_REG',
  input_label_cols = ['body_mass_g']
) AS
SELECT species, island, culmen_length_mm, flipper_length_mm, sex, body_mass_g
FROM `bigquery-public-data.ml_datasets.penguins`
WHERE body_mass_g IS NOT NULL AND sex IN ('MALE', 'FEMALE')
"""
client.query(query).result()
print('Model linear_regression_penguins_transform created')

Model linear_regression_penguins_transform created


In [16]:
# Predict on RAW rows - scaling is applied automatically inside the model
query = f"""
SELECT predicted_body_mass_g, culmen_length_mm, flipper_length_mm
FROM ML.PREDICT(
  MODEL `{PROJECT_ID}.{DATASET_ID}.linear_regression_penguins_transform`,
  (SELECT * FROM `bigquery-public-data.ml_datasets.penguins`
   WHERE body_mass_g IS NOT NULL AND sex IN ('MALE', 'FEMALE') LIMIT 5)
)
"""
client.query(query).to_dataframe()

,predicted_body_mass_g,culmen_length_mm,flipper_length_mm
0,3316.485482,36.6,184.0
1,3849.016623,39.8,184.0
2,3872.697834,40.9,184.0
3,3381.673418,46.5,192.0
4,3476.657826,37.3,192.0


---
## Step 9 — Hyperparameter tuning

BigQuery ML has built-in hyperparameter tuning. Set `num_trials` and define a search space with `HPARAM_RANGE` / `HPARAM_CANDIDATES`; BigQuery runs the trials and keeps the best model by `hparam_tuning_objectives`. Inspect every trial with `ML.TRIAL_INFO`.

Training multiple trials takes a few minutes.

In [17]:
query = f"""
CREATE OR REPLACE MODEL `{PROJECT_ID}.{DATASET_ID}.linear_regression_penguins_tuned`
OPTIONS(
  model_type = 'LINEAR_REG',
  input_label_cols = ['body_mass_g'],
  num_trials = 10,
  max_parallel_trials = 2,
  hparam_tuning_objectives = ['r2_score'],
  l1_reg = HPARAM_RANGE(0, 10),
  l2_reg = HPARAM_RANGE(0, 10)
) AS
SELECT
  species, island, culmen_length_mm, culmen_depth_mm, flipper_length_mm, sex, body_mass_g
FROM `bigquery-public-data.ml_datasets.penguins`
WHERE body_mass_g IS NOT NULL AND sex IN ('MALE', 'FEMALE')
"""
client.query(query).result()
print('Tuned model created')

Tuned model created


In [18]:
query = f"""
SELECT
  trial_id,
  hyperparameters,
  hparam_tuning_evaluation_metrics.r2_score AS r2_score,
  is_optimal
FROM ML.TRIAL_INFO(MODEL `{PROJECT_ID}.{DATASET_ID}.linear_regression_penguins_tuned`)
ORDER BY r2_score DESC
"""
client.query(query).to_dataframe()

,trial_id,hyperparameters,r2_score,is_optimal
0,4,"{'l1_reg': 9.942669313681121e-12, 'l2_reg': 0.0006295301896878236}",0.864325,True
1,8,"{'l1_reg': 8.128961175000337e-09, 'l2_reg': 0.0001285270840154018}",0.864325,False
2,10,"{'l1_reg': 1.2387219400241712e-10, 'l2_reg': 6.380072742819638e-05}",0.864325,False
3,7,"{'l1_reg': 1e-14, 'l2_reg': 6.043118097640522e-05}",0.864325,False
4,9,"{'l1_reg': 1.3253999341589917e-11, 'l2_reg': 6.811177813200311e-06}",0.864325,False
5,3,"{'l1_reg': 7.868027312253143e-08, 'l2_reg': 7.933981818051664e-08}",0.864325,False
6,2,"{'l1_reg': 1.981501369154032e-11, 'l2_reg': 1.9951356397597975e-11}",0.864325,False
7,1,"{'l1_reg': 1e-14, 'l2_reg': 1e-14}",0.864325,False
8,5,"{'l1_reg': 1e-14, 'l2_reg': 9.974059522679918}",0.847155,False
9,6,"{'l1_reg': 2.0160247496682176e-08, 'l2_reg': 9.99999999999999}",0.847020,False


---
## Examples — `%%bigquery` Magics

The same operations using IPython magic commands — write SQL directly in cells without Python string wrapping.

- `%%bigquery` — run SQL, display results inline
- `%%bigquery df` — run SQL, capture results into a pandas DataFrame

### Evaluate with `%%bigquery`

In [19]:
%%bigquery --project {PROJECT_ID}

SELECT *
FROM ML.EVALUATE(MODEL `statmike-mlops-349915.bq_ml.linear_regression_penguins`)

Query is running:   0%|          |

Downloading:   0%|          |

,mean_absolute_error,mean_squared_error,mean_squared_log_error,median_absolute_error,r2_score,explained_variance
0,226.413652,80658.955707,0.004982,173.534936,0.875223,0.875223


---
## Examples — BigFrames

BigFrames provides a scikit-learn-style API (`bigframes.ml`) that trains BigQuery ML models under the hood. Here's the same linear regression with `bigframes.ml.linear_model.LinearRegression`.

In [20]:
import bigframes.pandas as bpd

bpd.close_session()  # Reset session to apply project/location settings
bpd.options.bigquery.project = PROJECT_ID
bpd.options.bigquery.location = LOCATION

In [21]:
from bigframes.ml.linear_model import LinearRegression

# Load data as a BigFrames DataFrame
df = bpd.read_gbq('bigquery-public-data.ml_datasets.penguins')
df = df[df['body_mass_g'].notnull() & df['sex'].isin(['MALE', 'FEMALE'])]

feature_cols = ['species', 'island', 'culmen_length_mm', 'culmen_depth_mm', 'flipper_length_mm', 'sex']
X = df[feature_cols]
y = df['body_mass_g']

# Train (creates a BigQuery ML model behind the scenes)
model = LinearRegression()
model.fit(X, y)

# Evaluate
model.score(X, y).to_pandas()

,mean_absolute_error,mean_squared_error,mean_squared_log_error,median_absolute_error,r2_score,explained_variance
0,226.413652,80658.955707,0.004982,173.534936,0.875223,0.875223


---
## Cleanup

Drop the models created by this notebook. The shared dataset is left for other notebooks.

In [22]:
for model_name in [
    'linear_regression_penguins',
    'linear_regression_penguins_transform',
    'linear_regression_penguins_tuned',
]:
    client.query(f'DROP MODEL IF EXISTS `{PROJECT_ID}.{DATASET_ID}.{model_name}`').result()
    print(f'Dropped {model_name}')

Dropped linear_regression_penguins
Dropped linear_regression_penguins_transform
Dropped linear_regression_penguins_tuned


### Remove all shared resources

When finished with **all** notebooks, uncomment and run the cell below to delete the shared dataset and everything in it.

In [23]:
# client.delete_dataset(
#     f'{PROJECT_ID}.{DATASET_ID}',
#     delete_contents=True,
#     not_found_ok=True
# )
# print(f'Dataset {PROJECT_ID}.{DATASET_ID} deleted')